designs/v1alpha4-api.md

```diff
@@ -0,0 +1,115 @@
+```yaml
+spec:
+  zones: ["us-west-2a", "us-west-2b"] # Vendor neutral
+  securityGroups: ["abc", "def"] # Vendor specific, but sibling of zones
+```
+
+1. (+) Parameters that are natural siblings (e.g. `zones`, `securityGroups`) will be no longer be separated by `provider`.
+2. (-) It's not immediately clear to readers which parameters are vendor specific.
+3. (-) Key name conflicts may arise between provider and vendor fields.
+4. (-) `kind` and `apiVersion` are awkwardly placed at the top level of spec.
+
+### Alternative: Duck Typing
+
+It's possible to follow [Knative's Duck Typing approach](https://www.youtube.com/watch?v=kldVg63Utuw) and build vendor specific CRDs that contain vendor neutral snippets (or ducks) that Karpenter can recognize. Cloud Providers would consume Karpenter generic controllers as libaries, which would behave against the generic API snippets. For example:
+
+```yaml
+apiVersion: karpenter.k8s.aws/v1alpha1 # Vendor Specific
+kind: Provisioner
+spec:
+  limits: {} # Vendor neutral, recognized by Karpenter generic controllers
+  labels: {} # Vendor neutral, recognized by Karpenter generic controllers
+  taints: [] # Vendor netural, recognized by Karpenter generic controllers
+  subnets: [] # Vendor specific, only recognized by AWS Cloud Provider code
+```
+
+1. (+) Versioning is only defined in one place
+2. (+) Vendors may change versions decoupled from each other.
+3. (+) Parameters that are natural siblings (e.g. `zones`, `securityGroups`) will be no longer be separated by `provider`.
+4. (-) Cloud Provider must do more than simply implementing an interface (they must define an CR, etc).
+5. (-) Potential for name collision as duck types evolve or confusion about what fields providers own.
+6. (-) Knative duck typing APIs are alpha, missing documentation, and not widely adopted.
+
+## Pluralizion for all Constraints
+We expand the constraints of a Provisioner from a scalar to a vector in all cases. This will apply to both `operatingSystem` and `architecture` enabling the operator to specify greater flexibility. This supports use cases such as heterogenous architectures within a single provisioner, selected dynamically at runtime by the cloud provider. More importantly, this change creates a consistent and predictable semantic for all vendor neutral constraints. Similar to other constraints, the semantic of this change allows cloud providers to choose any value in the constraint slice, e.g. prioritizing arm64 for cost reasons.
+### Example
+```yaml
+spec:
+  operatingSystems: ["linux", "windows"] # operatingSystem -> operatingSystems
+  architectures: ["amd64", "arm64"] # architecture -> architectures
+```
+
+## Limits
+
+We introduce a new field `spec.limits` that contains configuration parameters to limit scaling and control costs.
+
+### Example
+```yaml
+spec:
+  limits:
+    ready: 500 # Flat or Percentage. Karpenter will not launch additional capacity if ready nodes exceeds this value
```

Comment:
What does "ready=90%" mean?  (percentage of what total?)

Reply:
This is a good point. It slipped my mind when I was thinking "IntOrString all the things" and "unready should have parity, so let's make ready". I suppose it could mean, this provisioner can deploy 50% of the nodes in this cluster. Though I'd much prefer that the provisioners are their own universes and are not aware of other provisioners. 

I keep going back and forth on a node count as a limit (as opposed to resource limits below). I'm leaning simply leaving this out entirely until I have a real use case for it.